R/preprocess.R

```diff
@@ -21,41 +21,60 @@ enw_metadata <- function(obs, target_date = "reference_date") {
   return(metaobs[])
 }
 
-#' @title FUNCTION_TITLE
-#' @description FUNCTION_DESCRIPTION
-#' @param metaobs PARAM_DESCRIPTION
-#' @param holidays PARAM_DESCRIPTION
+#' @title add common metadata variables
+#'
+#' @description annotates time series data with commonly required data
+#'
+#' @param metaobs raw meta data for observations, coerceable via [data.table::as.data.table()].
+#'  must have a `date` column
 #' @param holidays_to A character string to assign to holidays when present.
 #' Replaces the day of the week and defaults to Sunday.
-#' @return OUTPUT_DESCRIPTION
+#' @param holidays an optional vector of items coerceable via [data.table::as.IDate()]
+#'
+#' @return a copy of the `metaobs` input, with additional columns:
+#'  * `day_of_week`, a factor of values as output from [weekday()] and possibly as `holiday_to` if distinct from weekdays values
+#'  * `day`, numeric, 0 based from start of time series
+#'  * `week`, numeric, 0 based from start of time series
+#'  * `month`, numeric, 0 based from start of time series
+#'
 #' @family preprocess
 #' @export
 #' @importFrom data.table as.data.table
-enw_add_metaobs_features <- function(metaobs, holidays = c(),
-                                     holidays_to = "Sunday") {
-  # add days of week
+enw_add_metaobs_features <- function(
+  metaobs,
+  holidays_to = "Sunday",
+  holidays,
+  ...
+) {
   metaobs <- data.table::copy(metaobs)
-  metaobs[, day_of_week := weekdays(date)]
+  if (is.null(metaobs$date)) stop("`metaobs` does not have a column `date`.")
+
+  # add days of week
+  metaobs[, day_of_week := weekdays(date) ]
 
-  # make holidays be Sundays
-  if (length(holidays) != 0) {
-    metaobs[get(holidays) == TRUE, day_of_week := holidays_to]
+  # set holidays to associated day_of_week
+  if (!missing(holidays) && (length(holidays) > 0)) {
+    holidays <- as.IDate(holidays, optional = TRUE)
+    if (any(is.na(holidays))) stop(sprintf("Failed to parse holidays; first failure at index %i.", which.max(is.na(holidays))))
+    metaobs[date %in% holidays, day_of_week := holidays_to ]
   }
 
   # make day of week a factor
-  metaobs[, day_of_week := factor(day_of_week)]
-
-  # add day feature
-  metaobs[, day := as.numeric(date)]
-  metaobs[, day := day - min(day)]
+  metaobs[, day_of_week := factor(day_of_week) ]
 
-  # add week feature
-  metaobs[, week := lubridate::week(date)]
-  metaobs[, week := week - min(week)]
+  zerobase <- function(x) x - min(x)
 
-  # add month feature
-  metaobs[, month := lubridate::month(date)]
-  metaobs[, month := month - min(month)]
+  # add day / week / month features
+  metaobs[,
+    c("day", "week", "month") :=
+    lapply(
+      lapply(
+        .(as.numeric, lubridate::week, lubridate::month),
```

Comment:
n.b. this now causes warnings if e.g. data.table loaded (since it exports week and month).

bit unclear on why `importFrom`ing here if going to access via `::`

Reply:
pretty much always do both to make sure the namespace is a complete list but you are right it should be functionally equiv. Do the `data.table` variants work in the same way? Perhaps we should move to those and drop `lubricate`?